File: Simulatron/simulatron.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void print(short instruction, int accumulator, int * memory, short mem_length);

int main(int argc, char * argv[]) {
	if(argc < 2 || argc > 3) {
		fprintf(stderr, "Usage: %s [-d] <filename>\n", argv[0]);
		return 1;
	}

	short debug;
	if(argc == 3 && strcmp(argv[2], "-d") == 0)
		debug = 1;
	else
		debug = 0;

	FILE * program = fopen(argv[argc - 1], "r");
	if(program == NULL) {
		perror("Could not open program");
		return 2;
	}

	short warnings = debug;
	short mem_length = 100, mem_width, val_width, instruction = 0;
	int accumulator = 0;

	char option[20];
	int result = fscanf(program, "#%20s\n", option);
	while(result) {
		if(sscanf(option, "memory %hd", &mem_length) == 1);
		else if(!warnings && sscanf(option, "warnings %1hd", &warnings) == 1);
		else {
			fprintf(stderr, "Invalid pragma found: %s", option);
			return 5;
		}

		result = fscanf(program, "#%20s\n", option);
	}

	if(mem_length < 0 || mem_length > 32767) {
		fprintf(stderr, "Error: Invalid memory size\n");
		return 1;
	}

	mem_width = pow(10, ceil(log10(mem_length)));
	val_width = mem_width * mem_width;

	int * memory = calloc(mem_length, sizeof(int));

	short i = 0;
	do {
		if(i >= mem_length) {
			fprintf(stderr, "Error: Not enough memory to load program\n");
			free(memory);
			fclose(program);
			return 3;
		}

		result = fscanf(program, "%d", &(memory[i]));
		i++;
	}
	while(result > 0 && result != EOF);

	fclose(program);

	short error = 0;
	short debug_prompt = debug;
	short breakpoint = -1;
	while(1) {
		while(debug_prompt || error || instruction == breakpoint) {
			char command;
			char args[10];

			printf("(dgb) ");
			short num_args = scanf("%1s %10s", &command, args);
			if(num_args == 0)
				continue;

			if(command == 'r') {
				debug_prompt = 0;
				break;
			}
			else if(command == 'p')
				print(instruction, accumulator, memory, mem_length);
			else if(command == 's')
				break;
			else if(command == 'b') {
				if(!sscanf(args, "%hd", &breakpoint))
					printf("Usage: b <instruction>\n");
			}
			else if(command == 'q')
				return 0;
			else if(command == 'h')
				printf("r - Run program from beginning or continue from current position.\np - Print the instruction pointer, instruction register, accumulator, and memory\ns - Step one instruction\nb - Set a breakpoint\nq - Quit\nh - Display this help\n");
			else
				printf("Unknown command");
		}

		if(error) {
			free(memory);
			return error;
		}

		if(instruction >= mem_length) {
			fprintf(stderr, "Error: Hit end of memory space\n");
			if(debug) {
				error = 5;
				continue;
			}
			free(memory);
			return 5;
		}

		if(memory[instruction] == 0) {
			if(warnings)
				fprintf(stderr, "Warning: Executing uninitialized space at %d\n", instruction);
			while(memory[instruction] == 0) {
				instruction++;
				if(instruction >= mem_length) {
					fprintf(stderr, "Error: Hit end of memory space\n");
					if(debug) {
						error = 5;
						continue;
					}
					free(memory);
					return 5;
				}
			}
		}

		int opcode = memory[instruction] / mem_width;
		int argument = memory[instruction] % mem_width;

		if(argument < 0 || argument >= mem_length) {
			fprintf(stderr, "Error: Invalid argument %d at %d\n", argument, instruction);
			if(debug) {
				error = 4;
				continue;
			}
			free(memory);
			return 4;
		}

		switch(opcode) {
			case 10:
				printf("> ");
				if(!scanf("%d", &(memory[argument]))) {
					fprintf(stderr, "Error: Input not a number\n");
					if(debug) {
						error = 6;
						continue;
					}
					free(memory);
					return 6;
				}
				break;
			case 11:
				printf("%d\n", memory[argument]);
				break;
			case 20:
				accumulator = memory[argument];
				if(warnings && (accumulator >= val_width || accumulator <= -val_width))
					fprintf(stderr, "Warning: Accumulator overflow at %d\n", instruction);
				break;
			case 21:
				memory[argument] = accumulator;
				break;
			case 30:
				accumulator += memory[argument];
				if(warnings && (accumulator >= val_width || accumulator <= -val_width))
					fprintf(stderr, "Warning: Accumulator overflow at %d\n", instruction);
				break;
			case 31:
				accumulator -= memory[argument];
				if(warnings && (accumulator >= val_width || accumulator <= -val_width))
					fprintf(stderr, "Warning: Accumulator overflow at %d\n", instruction);
				break;
			case 32:
				if(memory[argument] == 0) {
					fprintf(stderr, "Error: Division by zero at %d\n", instruction);
					if(debug) {
						error = 4;
						continue;
					}
					free(memory);
					return 4;
				}
				accumulator /= memory[argument];
				if(warnings && (accumulator >= val_width || accumulator <= -val_width))
					fprintf(stderr, "Warning: Accumulator overflow at %d\n", instruction);
				break;
			case 33:
				accumulator *= memory[argument];
				if(warnings && (accumulator >= val_width || accumulator <= -val_width))
					fprintf(stderr, "Warning: Accumulator overflow at %d\n", instruction);
				break;
			case 40:
				instruction = argument;
				continue;
			case 41:
				if(accumulator < 0)
					instruction = argument;
				break;
			case 42:
				if(accumulator == 0)
					instruction = argument;
				break;
			case 43:
				free(memory);
				return 0;
			case 50:
				break;
			default:
				fprintf(stderr, "Error: Unknown opcode %d at %d\n", opcode, instruction);
				if(debug) {
					error = 4;
					continue;
				}
				free(memory);
				return 4;
		}

		instruction++;
	}
}

void print(short instruction, int accumulator, int * memory, short mem_length) {
	fprintf(stderr, "Instruction Pointer: %hd\n", instruction);
	fprintf(stderr, "Instruction Register: %d\n", memory[instruction]);

	fprintf(stderr, "Accumulator: %d\n", accumulator);

	int number_width = ceil(log10(mem_length));
	fprintf(stderr, "Memory:\n");
	for(int i = 0; i < mem_length; i++) {
		fprintf(stderr, "%*d |", i, number_width);
		for(int ii = 0; ii < 10; ii++, i++)
			fprintf(stderr, " %d", memory[i]);
		fprintf(stderr, "\n");
	}
}

File: Simulatron/java/Simulatron.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.InputMismatchException;
import java.util.Scanner;
import javax.swing.JFileChooser;
import javax.swing.JOptionPane;

public class Simulatron {
	public static void main(String[] args) {
		JFileChooser chooser = new JFileChooser();
		File file = new File("");

		if(chooser.showOpenDialog(null) == JFileChooser.APPROVE_OPTION)
			file = chooser.getSelectedFile();

		Scanner scanner;
		try {
			scanner = new Scanner(file);
		}
		catch(FileNotFoundException e) {
			System.err.println("Error: File not found");
			return;
		}

		boolean debug = JOptionPane.showConfirmDialog(null, "Would you like to enable debugging?", "Simulatron", JOptionPane.YES_NO_OPTION) == JOptionPane.YES_OPTION;
		boolean debug_prompt = debug;

		int[] memory = new int[100];
		int i = 0;

		try {
			//Check the entire file for numbers
			while(scanner.hasNext()) {
				try {
					int value = Integer.parseInt(scanner.next());
					if(value > 9999 || value < -9999) { //Check it's length
						System.err.println("Error: Memory value " + value + " out of range at " + i);
						return;
					}
					memory[i] = value;
					i++;
				}
				catch(NumberFormatException e) {} //Ignore non-numerical tokens as comments
			}
		}
		catch(ArrayIndexOutOfBoundsException e) {
			System.err.println("Error: Not enough memory to load program");
			return;
		}

		Simulator simulator = new Simulator(memory);

		boolean error = false;
		int breakpoint = -1;
		boolean run = true;
		while(run) {
			try {
				//Check if we hit a breakpoint
				boolean hit_breakpoint = false;
				if(debug && simulator.getInstruction() == breakpoint) {
					hit_breakpoint = true;
					System.err.println("Hit breakpoint at " + simulator.getInstruction());
				}

				//Check if a debug prompt is necessary
				while(debug && (debug_prompt || error || hit_breakpoint)) {
					System.err.print("(dbg) ");
					//Grab the command each argument
					Scanner stdin = new Scanner(System.in);
					String[] command = stdin.nextLine().split(" ");

					//Restart from beginning
					if(command[0].equalsIgnoreCase("r") || command[0].equalsIgnoreCase("run")) {
						error = false;
						debug_prompt = false;
						simulator.setInstruction(0);
						break;
					}
					//Just turn off prompt and break out of debug loop
					else if(command[0].equalsIgnoreCase("c") || command[0].equalsIgnoreCase("continue")) {
						debug_prompt = false;
						break;
					}
					else if(command[0].equalsIgnoreCase("p") || command[0].equalsIgnoreCase("print")) {
						try {
							if(command.length > 2)
								throw new Exception();
							if(command.length == 2) {
								int n = Integer.parseInt(command[1]);
								if(n > 99 || n < 0) {
									System.err.println("Error: n out of range");
									continue;
								}
								simulator.print(n);
							}
							else {
								simulator.print();
							}
						}
						//Some type of argument failure
						catch(Exception e) {
							System.err.println("Usage: p(rint) [n]");
						}
					}
					else if(command[0].equalsIgnoreCase("w") || command[0].equalsIgnoreCase("write")) {
						try {
							if(command.length > 3)
								throw new Exception();
							int n = Integer.parseInt(command[1]);
							if(n > 99 || n < 0) {
								System.err.println("Error: n out of range");
								continue;
							}
							int val = Integer.parseInt(command[2]);
							if(val > 9999 || val < -9999) {
								System.err.println("Error: val out of range");
								continue;
							}
							simulator.set(n, val);
						}
						catch(Exception e) {
							System.err.println("Usage: w(rite) <n> <val>");
						}
					}
					else if(command[0].equalsIgnoreCase("s") || command[0].equalsIgnoreCase("step")) {
						debug_prompt = true;
						break;
					}
					else if(command[0].equalsIgnoreCase("b") || command[0].equalsIgnoreCase("break")) {
						try {
							if(command.length > 2)
								throw new Exception();
							if(command.length == 2) {
								int n = Integer.parseInt(command[1]);
								if(n > 99 || n < 0) {
									System.err.println("Error: n out of range");
									continue;
								}
								breakpoint = n;
							}
							else {
								breakpoint = 0;
								System.err.println("Breakpoint cleared");
							}
						}
						catch(Exception e) {
							System.err.println("Usage: b(reak) [n]");
						}
					}
					else if(command[0].equalsIgnoreCase("q") || command[0].equalsIgnoreCase("quit")) {
						return;
					}
					else if(command[0].equalsIgnoreCase("h") || command[0].equalsIgnoreCase("help")) {
						System.err.println("r(un)		Run program from beginning");
						System.err.println("c(ontinue)		Continue program from current position");
						System.err.println("p(rint) [n]		Print the instruction pointer, instruction register, accumulator, and memory or, if specified, print the memory space n");
						System.err.println("w(rite) <n> <val>	Write val to memory space n");
						System.err.println("s(tep)		Step one instruction");
						System.err.println("b(reak) [n]		Set a breakpoint at memory space n or if n is not specified, clear the breakpoint");
						System.err.println("q(uit)		Quit");
						System.err.println("h(elp)		Display this help");
					}
					else {
						System.err.println("Unknown command");
					}
				}

				//If we have a standing error, exit
				if(error) {
					run = false;
					continue;
				}

				if(!simulator.step()) {
					if(debug)
						error = true; //End with a debug prompt if in debug mode
					else
						run = false;
				}
			}
			catch(OutOfMemoryException e) {
				System.err.println("Error: Hit end of memory while executing");
				error = true;
			}
			catch(DivisionByZeroException e) {
				System.err.println("Error: Attempt to divide by zero at " + simulator.getInstruction());
				error = true;
			}
			catch(InvalidOpcodeException e) {
				System.err.println("Error: Invalid opcode " + simulator.get(simulator.getInstruction()) + " at " + simulator.getInstruction());
				error = true;
			}
			catch(AccumulatorOverflowException e) {
				System.err.println("Error: Accumulator overflow at " + simulator.getInstruction());
				error = true;
			}
			catch(InputMismatchException e) {
				System.err.println("Error: Input not a number");
				error = true;
			}
		}
	}
}
